File: app/src/main/java/de/westnordost/streetcomplete/quests/validator/multidesignatedFootwayToPath.java
package de.westnordost.streetcomplete.quests.validator;

import android.os.Bundle;
import android.support.annotation.NonNull;

import java.util.Map;

import javax.inject.Inject;

import de.westnordost.streetcomplete.R;
import de.westnordost.streetcomplete.data.osm.SimpleOverpassQuestType;
import de.westnordost.streetcomplete.data.osm.changes.StringMapChangesBuilder;
import de.westnordost.streetcomplete.data.osm.download.OverpassMapDataDao;
import de.westnordost.streetcomplete.quests.AbstractQuestAnswerFragment;

public class multidesignatedFootwayToPath extends SimpleOverpassQuestType
{
	@Inject public multidesignatedFootwayToPath(OverpassMapDataDao overpassServer)
	{
		super(overpassServer);
	}

	@Override
	protected String getTagFilters()
	{
		return "ways with highway=footway and bicycle=designated and (foot=designated or !foot)";
	}

	public AbstractQuestAnswerFragment createForm()
	{
		AbstractQuestAnswerFragment form = new multidesignatedFootwayToPathForm();
		return form;
	}

	public void applyAnswerTo(Bundle answer, StringMapChangesBuilder changes)
	{
		if(!answer.getBoolean(multidesignatedFootwayToPathForm.ANSWER))
		{
			//TODO: handle nonexisting pedestrian + cyclist route marked on the map
			return;
		}
		changes.addOrModify("foot", "designated");
		changes.modify("highway", "path");
	}

	@Override public String getCommitMessage()
	{
		return "fix misused highway=footway, confirmation that route for both pedestrian and cyclists exists";
	}

	@Override
	public int getTitle(@NonNull Map<String, String> tags) {
		return R.string.quest_multidesignatedFootway_title;
	}

	@Override public int getIcon() { return R.drawable.ic_quest_bicycle; }}

File: app/src/main/java/de/westnordost/streetcomplete/quests/validator/MultidesignatedFootwayToPath.kt
package de.westnordost.streetcomplete.quests.validator

import de.westnordost.streetcomplete.R
import de.westnordost.streetcomplete.data.osm.edits.update_tags.StringMapChangesBuilder
import de.westnordost.streetcomplete.data.osm.geometry.ElementGeometry
import de.westnordost.streetcomplete.data.osm.osmquests.OsmFilterQuestType
import de.westnordost.streetcomplete.data.user.achievements.EditTypeAchievement
import de.westnordost.streetcomplete.osm.Tags
import de.westnordost.streetcomplete.quests.YesNoQuestForm

class MultidesignatedFootwayToPath() : OsmFilterQuestType<Boolean>() {

    override val elementFilter = "ways with highway=footway and bicycle=designated and (foot=designated or !foot)"
    override val changesetComment = "fix misused highway=footway, confirmation that route for both pedestrian and cyclists exists"
    override val icon = R.drawable.ic_quest_bicycle

    override fun getTitle(tags: Map<String, String>) = R.string.quest_multidesignatedFootway_title

    override fun createForm() = YesNoQuestForm()

    override fun applyAnswerTo(answer: Boolean, tags: Tags, geometry: ElementGeometry, timestampEdited: Long) {
        if (!answer) {
            //TODO: handle nonexisting pedestrian + cyclist route marked on the map
            return
        }
        tags["foot"] = "designated"
        tags["highway"] = "path"
    }

    override val wikiLink = "Tag:highway=path"

    override val achievements: List<EditTypeAchievement> = listOf()
}
